File: backend/app/schemas/novel.py
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ChoiceOption(BaseModel):
    """前端选择项描述，用于动态 UI 控件。"""

    id: str
    label: str


class UIControl(BaseModel):
    """描述前端应渲染的组件类型与配置。"""

    type: str = Field(..., description="控件类型，如 single_choice/text_input")
    options: Optional[List[ChoiceOption]] = Field(default=None, description="可选项列表")
    placeholder: Optional[str] = Field(default=None, description="输入提示文案")


class ConverseResponse(BaseModel):
    """概念对话接口的统一返回体。"""

    ai_message: str
    ui_control: UIControl
    conversation_state: Dict[str, Any]
    is_complete: bool = False
    ready_for_blueprint: Optional[bool] = None


class ConverseRequest(BaseModel):
    """概念对话接口的请求体。"""

    user_input: Dict[str, Any]
    conversation_state: Dict[str, Any]


class ChapterGenerationStatus(str, Enum):
    NOT_GENERATED = "not_generated"
    GENERATING = "generating"
    EVALUATING = "evaluating"
    SELECTING = "selecting"
    FAILED = "failed"
    EVALUATION_FAILED = "evaluation_failed"
    WAITING_FOR_CONFIRM = "waiting_for_confirm"
    SUCCESSFUL = "successful"


class ChapterOutline(BaseModel):
    chapter_number: int
    title: str
    summary: str


class PartOutlineStatus(str, Enum):
    """部分大纲生成状态"""
    PENDING = "pending"
    GENERATING = "generating"
    COMPLETED = "completed"
    FAILED = "failed"


class PartOutline(BaseModel):
    """部分大纲（用于长篇小说的分层结构）"""
    part_number: int
    title: str
    start_chapter: int
    end_chapter: int
    summary: str
    theme: str
    key_events: List[str] = []
    character_arcs: Dict[str, str] = {}  # 角色名 -> 成长描述
    conflicts: List[str] = []
    ending_hook: Optional[str] = None  # 与下一部分的衔接点
    generation_status: PartOutlineStatus = PartOutlineStatus.PENDING
    progress: int = 0  # 生成进度 0-100


class Chapter(ChapterOutline):
    real_summary: Optional[str] = None
    content: Optional[str] = None
    versions: Optional[List[str]] = None
    evaluation: Optional[str] = None
    generation_status: ChapterGenerationStatus = ChapterGenerationStatus.NOT_GENERATED


class Relationship(BaseModel):
    character_from: str
    character_to: str
    description: str


class Blueprint(BaseModel):
    title: str
    target_audience: str = ""
    genre: str = ""
    style: str = ""
    tone: str = ""
    one_sentence_summary: str = ""
    full_synopsis: str = ""
    world_setting: Dict[str, Any] = {}
    characters: List[Dict[str, Any]] = []
    relationships: List[Relationship] = []
    chapter_outline: List[ChapterOutline] = []
    needs_part_outlines: bool = False  # 是否需要分阶段生成（>50章）
    total_chapters: Optional[int] = None  # 总章节数
    chapters_per_part: Optional[int] = 25  # 每部分章节数（默认25）
    part_outlines: List[PartOutline] = []  # 部分大纲列表


class NovelProject(BaseModel):
    id: str
    user_id: int
    title: str
    initial_prompt: str
    status: str  # 项目状态：draft, blueprint_ready, part_outlines_ready, chapter_outlines_ready 等
    conversation_history: List[Dict[str, Any]] = []
    blueprint: Optional[Blueprint] = None
    chapters: List[Chapter] = []

    class Config:
        from_attributes = True


class NovelProjectSummary(BaseModel):
    id: str
    title: str
    genre: str
    last_edited: str
    completed_chapters: int
    total_chapters: int
    status: str  # 项目状态：draft, blueprint_ready, part_outlines_ready, chapter_outlines_ready 等


class BlueprintGenerationResponse(BaseModel):
    blueprint: Blueprint
    ai_message: str


class BlueprintRefineRequest(BaseModel):
    """蓝图优化请求体"""
    refinement_instruction: str = Field(
        ...,
        description="用户的优化指令，描述想要改进的方向",
        min_length=1,
        max_length=2000
    )


class ChapterGenerationResponse(BaseModel):
    ai_message: str
    chapter_versions: List[Dict[str, Any]]


class NovelSectionType(str, Enum):
    OVERVIEW = "overview"
    WORLD_SETTING = "world_setting"
    CHARACTERS = "characters"
    RELATIONSHIPS = "relationships"
    CHAPTER_OUTLINE = "chapter_outline"
    CHAPTERS = "chapters"


class NovelSectionResponse(BaseModel):
    section: NovelSectionType
    data: Dict[str, Any]


class GenerateChapterRequest(BaseModel):
    chapter_number: int
    writing_notes: Optional[str] = Field(default=None, description="章节额外写作指令")


class SelectVersionRequest(BaseModel):
    chapter_number: int
    version_index: int


class EvaluateChapterRequest(BaseModel):
    chapter_number: int


class UpdateChapterOutlineRequest(BaseModel):
    chapter_number: int
    title: str
    summary: str


class DeleteChapterRequest(BaseModel):
    chapter_numbers: List[int]


class GenerateOutlineRequest(BaseModel):
    start_chapter: int
    num_chapters: int


class BlueprintPatch(BaseModel):
    one_sentence_summary: Optional[str] = None
    full_synopsis: Optional[str] = None
    world_setting: Optional[Dict[str, Any]] = None
    characters: Optional[List[Dict[str, Any]]] = None
    relationships: Optional[List[Relationship]] = None
    chapter_outline: Optional[List[ChapterOutline]] = None


class EditChapterRequest(BaseModel):
    chapter_number: int
    content: str


# 部分大纲相关请求和响应模型
class GeneratePartOutlinesRequest(BaseModel):
    """生成部分大纲请求"""
    total_chapters: int = Field(
        ...,
        description="小说总章节数",
        ge=10,
        le=10000
    )
    chapters_per_part: int = Field(
        default=25,
        description="每个部分的章节数",
        ge=10,
        le=100
    )


class GeneratePartChaptersRequest(BaseModel):
    """基于部分大纲生成章节请求"""
    regenerate: bool = Field(
        default=False,
        description="是否重新生成（如果章节已存在则覆盖）"
    )


class BatchGenerateChaptersRequest(BaseModel):
    """批量并发生成章节请求"""
    part_numbers: Optional[List[int]] = Field(
        default=None,
        description="要并发生成的部分编号列表，为空则生成所有"
    )
    max_concurrent: int = Field(
        default=5,
        description="最大并发数",
        ge=1,
        le=10
    )


class PartOutlineGenerationProgress(BaseModel):
    """部分大纲批量生成进度（整体进度）"""
    parts: List[PartOutline] = []  # 所有部分的大纲列表
    total_parts: int = 0  # 总部分数
    completed_parts: int = 0  # 已完成部分数
    status: str = "pending"  # 整体状态：pending, partial, completed

File: frontend/src/api/novel.ts
import { useAuthStore } from '@/stores/auth'
import router from '@/router'

// API 配置
// 在生产环境中使用相对路径，在开发环境中使用绝对路径
export const API_BASE_URL = import.meta.env.MODE === 'production' ? '' : 'http://localhost:8000'
export const API_PREFIX = '/api'

// 统一的请求处理函数
const request = async (url: string, options: RequestInit = {}) => {
  const authStore = useAuthStore()
  const headers = new Headers({
    'Content-Type': 'application/json',
    ...options.headers
  })

  if (authStore.isAuthenticated && authStore.token) {
    headers.set('Authorization', `Bearer ${authStore.token}`)
  }

  const response = await fetch(url, { ...options, headers })

  if (response.status === 401) {
    // Token 失效或未授权
    authStore.logout()
    router.push('/login')
    throw new Error('会话已过期，请重新登录')
  }

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new Error(errorData.detail || `请求失败，状态码: ${response.status}`)
  }

  return response.json()
}

// 类型定义
export interface NovelProject {
  id: string
  title: string
  initial_prompt: string
  status: string  // 项目状态：draft, blueprint_ready, part_outlines_ready, chapter_outlines_ready 等
  blueprint?: Blueprint
  chapters: Chapter[]
  conversation_history: ConversationMessage[]
}

export interface NovelProjectSummary {
  id: string
  title: string
  genre: string
  last_edited: string
  completed_chapters: number
  total_chapters: number
  status: string  // 项目状态：draft, blueprint_ready, part_outlines_ready, chapter_outlines_ready 等
}

export interface Blueprint {
  title?: string
  target_audience?: string
  genre?: string
  style?: string
  tone?: string
  one_sentence_summary?: string
  full_synopsis?: string
  world_setting?: any
  characters?: Character[]
  relationships?: any[]
  chapter_outline?: ChapterOutline[]
  needs_part_outlines?: boolean  // 是否需要分阶段生成
  total_chapters?: number  // 总章节数
  chapters_per_part?: number  // 每部分章节数
  part_outlines?: PartOutline[]  // 部分大纲列表
}

export interface Character {
  name: string
  description: string
  identity?: string
  personality?: string
  goals?: string
  abilities?: string
  relationship_to_protagonist?: string
}

export interface ChapterOutline {
  chapter_number: number
  title: string
  summary: string
}

export interface PartOutline {
  part_number: number
  title: string
  start_chapter: number
  end_chapter: number
  summary: string
  theme: string
  key_events: string[]
  character_arcs: Record<string, string>
  conflicts: string[]
  ending_hook?: string
  generation_status: 'pending' | 'generating' | 'completed' | 'failed'
  progress: number
}

export interface PartOutlineProgress {
  parts: PartOutline[]
  total_parts: number
  completed_parts: number
  status: 'pending' | 'in_progress' | 'completed' | 'partial'
}

export interface ChapterVersion {
  content: string
  style?: string
}

export interface Chapter {
  chapter_number: number
  title: string
  summary: string
  content: string | null
  versions: string[] | null  // versions是字符串数组，不是对象数组
  evaluation: string | null
  generation_status: 'not_generated' | 'generating' | 'evaluating' | 'selecting' | 'failed' | 'evaluation_failed' | 'waiting_for_confirm' | 'successful'
  word_count?: number  // 字数统计
}

export interface ConversationMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface ConverseResponse {
  ai_message: string
  ui_control: UIControl
  conversation_state: any
  is_complete: boolean
  ready_for_blueprint?: boolean  // 新增：表示准备生成蓝图
}

export interface BlueprintGenerationResponse {
  blueprint: Blueprint
  ai_message: string
}

export interface UIControl {
  type: 'single_choice' | 'text_input'
  options?: Array<{ id: string; label: string }>
  placeholder?: string
}

export interface ChapterGenerationResponse {
  versions: ChapterVersion[] // Renamed from chapter_versions for consistency
  evaluation: string | null
  ai_message: string
  chapter_number: number
}

export interface DeleteNovelsResponse {
  status: string
  message: string
}

export type NovelSectionType = 'overview' | 'world_setting' | 'characters' | 'relationships' | 'chapter_outline' | 'chapters'

export interface NovelSectionResponse {
  section: NovelSectionType
  data: Record<string, any>
}

// API 函数
const NOVELS_BASE = `${API_BASE_URL}${API_PREFIX}/novels`
const WRITER_PREFIX = '/api/writer'
const WRITER_BASE = `${API_BASE_URL}${WRITER_PREFIX}/novels`

export class NovelAPI {
  static async createNovel(title: string, initialPrompt: string): Promise<NovelProject> {
    return request(NOVELS_BASE, {
      method: 'POST',
      body: JSON.stringify({ title, initial_prompt: initialPrompt })
    })
  }

  static async getNovel(projectId: string): Promise<NovelProject> {
    return request(`${NOVELS_BASE}/${projectId}`)
  }

  static async getChapter(projectId: string, chapterNumber: number): Promise<Chapter> {
    return request(`${NOVELS_BASE}/${projectId}/chapters/${chapterNumber}`)
  }

  static async getSection(projectId: string, section: NovelSectionType): Promise<NovelSectionResponse> {
    return request(`${NOVELS_BASE}/${projectId}/sections/${section}`)
  }

  static async converseConcept(
    projectId: string,
    userInput: any,
    conversationState: any = {}
  ): Promise<ConverseResponse> {
    const formattedUserInput = userInput || { id: null, value: null }
    return request(`${NOVELS_BASE}/${projectId}/concept/converse`, {
      method: 'POST',
      body: JSON.stringify({
        user_input: formattedUserInput,
        conversation_state: conversationState
      })
    })
  }

  static async generateBlueprint(projectId: string): Promise<BlueprintGenerationResponse> {
    return request(`${NOVELS_BASE}/${projectId}/blueprint/generate`, {
      method: 'POST'
    })
  }

  static async saveBlueprint(projectId: string, blueprint: Blueprint): Promise<NovelProject> {
    return request(`${NOVELS_BASE}/${projectId}/blueprint/save`, {
      method: 'POST',
      body: JSON.stringify(blueprint)
    })
  }

  static async refineBlueprint(projectId: string, refinementInstruction: string): Promise<BlueprintGenerationResponse> {
    return request(`${NOVELS_BASE}/${projectId}/blueprint/refine`, {
      method: 'POST',
      body: JSON.stringify({ refinement_instruction: refinementInstruction })
    })
  }

  static async generateChapter(projectId: string, chapterNumber: number): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/chapters/generate`, {
      method: 'POST',
      body: JSON.stringify({ chapter_number: chapterNumber })
    })
  }

  static async evaluateChapter(projectId: string, chapterNumber: number): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/chapters/evaluate`, {
      method: 'POST',
      body: JSON.stringify({ chapter_number: chapterNumber })
    })
  }

  static async selectChapterVersion(
    projectId: string,
    chapterNumber: number,
    versionIndex: number
  ): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/chapters/select`, {
      method: 'POST',
      body: JSON.stringify({
        chapter_number: chapterNumber,
        version_index: versionIndex
      })
    })
  }

  static async getAllNovels(): Promise<NovelProjectSummary[]> {
    return request(NOVELS_BASE)
  }

  static async deleteNovels(projectIds: string[]): Promise<DeleteNovelsResponse> {
    return request(NOVELS_BASE, {
      method: 'DELETE',
      body: JSON.stringify(projectIds)
    })
  }

  static async updateChapterOutline(
    projectId: string,
    chapterOutline: ChapterOutline
  ): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/chapters/update-outline`, {
      method: 'POST',
      body: JSON.stringify(chapterOutline)
    })
  }

  static async deleteChapter(
    projectId: string,
    chapterNumbers: number[]
  ): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/chapters/delete`, {
      method: 'POST',
      body: JSON.stringify({ chapter_numbers: chapterNumbers })
    })
  }

  static async generateChapterOutline(
    projectId: string,
    startChapter: number,
    numChapters: number
  ): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/chapters/outline`, {
      method: 'POST',
      body: JSON.stringify({
        start_chapter: startChapter,
        num_chapters: numChapters
      })
    })
  }

  // 短篇小说（≤50章）一次性生成全部章节大纲
  static async generateAllChapterOutlines(projectId: string): Promise<{
    message: string
    total_chapters: number
    status: string
  }> {
    return request(`${NOVELS_BASE}/${projectId}/chapter-outlines/generate`, {
      method: 'POST'
    })
  }

  static async updateBlueprint(projectId: string, data: Record<string, any>): Promise<NovelProject> {
    return request(`${NOVELS_BASE}/${projectId}/blueprint`, {
      method: 'PATCH',
      body: JSON.stringify(data)
    })
  }

  static async editChapterContent(
    projectId: string,
    chapterNumber: number,
    content: string
  ): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/chapters/edit`, {
      method: 'POST',
      body: JSON.stringify({
        chapter_number: chapterNumber,
        content: content
      })
    })
  }

  // 部分大纲相关API
  static async generatePartOutlines(
    projectId: string,
    totalChapters: number,
    chaptersPerPart: number = 25
  ): Promise<PartOutlineProgress> {
    return request(`${WRITER_BASE}/${projectId}/parts/generate`, {
      method: 'POST',
      body: JSON.stringify({
        total_chapters: totalChapters,
        chapters_per_part: chaptersPerPart
      })
    })
  }

  static async generateSinglePartChapters(
    projectId: string,
    partNumber: number,
    regenerate: boolean = false
  ): Promise<NovelProject> {
    return request(`${WRITER_BASE}/${projectId}/parts/${partNumber}/chapters`, {
      method: 'POST',
      body: JSON.stringify({
        regenerate: regenerate
      })
    })
  }

  static async getPartOutlinesProgress(projectId: string): Promise<PartOutlineProgress> {
    return request(`${WRITER_BASE}/${projectId}/parts/progress`)
  }

  static async batchGeneratePartChapters(
    projectId: string,
    partNumbers?: number[],
    maxConcurrent: number = 3
  ): Promise<PartOutlineProgress> {
    return request(`${WRITER_BASE}/${projectId}/parts/batch-generate`, {
      method: 'POST',
      body: JSON.stringify({
        part_numbers: partNumbers,
        max_concurrent: maxConcurrent
      })
    })
  }
}
